### PR TITLE
Add voting learners

### LINKED ISSUE #488: Add VotingClassifier and VotingRegressor ensembles.
These would be great for baseline ensembles.

### LINKED ISSUE #524: Use latest version of Python 3 for testing
At some point in the future, we we should remove the explicit dependency on Python 3.6 as the version of Python 3 we test on. We should just use whatever the latest version is.

### PR DESCRIPTION
**Pull request by desilinguist**
This PR mainly closes #488. It also closes #524 and #584. 

### New `VotingLearner` class

*Relevant files*: ``learner/voting.py`` and ``learner/utils.py``. 

- The main contribution of this PR is to allow SKLL to use `VotingClassifier` and `VotingRegressor` learners from scikit-learn. This was not as straightforward since this is the first true meta-learner class we have added to SKLL. By meta-learner, I mean a class that builds on top of SKLL's `Learner` class. This meta-learner class is called `VotingLearner``. 

- The implementation of `VotingLearner` provides the same 7 methods as in the original `Learner` class: `train()`, `cross_validate()`, `evaluate()`, `predict()` and `learning_curve()`, `from_file()` and `save()`. There is some minor code duplication between these methods and the corresponding methods in the `Learner` class but most of it was avoided by the refactoring of the common code into utility functions last year. There are some changes to those functions in order to accommodate the new meta-learner. In addition, there is some new refactoring primarily for the `from_file()` and `save()` methods for the two classes which now use the refactored functions `_save_learner_to_disk()` and `_load_learner_from_disk()`. 

- The implementation supports passing in keyword arguments to the underlying learners, as well as samplers and sampler arguments. 

- The `train()` implementation nicely incorporates grid search such that the learners underlying the voting meta-learner are automatically tuned with grid search (assuming the user requests it) before their predictions are used for voting. Grid search works with `cross_validate()` although, as expected, doing this is much slower since the per-fold-grid-search will now be done for *each* underlying learner. Both these methods also accept a list of parameter grids for tuning the underlying learners. 

- The `evaluate()` and `predict()` methods support returning (and writing out) not only the meta-learner's predictions but also the predictions from the underlying learners that were used in the voting process.

### Integration with `run_experiment`

*Relevant files*: ``config/__init__.py``, ``experiments/__init__.py``, and ``experiments/utils.py``.

- The `VotingLearner` class can be used via the SKLL API as described in the previous section. This PR also includes hooks that allow users to specify `VotingClassifier` and `VotingRegressor` as their chosen learners in an experiment configuration file as input to `run_experiment`. The hooks are set up to require users to specify the underlying estimators as fixed parameters using the `estimator_names` key. The following additional fixed parameters can also be specified: `voting_type`, `estimator_fixed_parameters`, `estimator_samplers`, `estimator_sampler_parameters`, and `estimator_param_grids`. These parameters are fully documented and example configuration files are also included (see documentation section below for details). 

- A new configuration field called `save_votes` is added to allow the user to save the predictions from the underlying learners in addition to the predictions from the `VotingClassifier` or `VotingRegressor`. The default value for this field is ``False``. 

- A new JSON encoder called `PipelineTypeEncoder()` was added so as to support being able to serialize `VotingLearner` instances to JSON. This was necessary since these instances in turn contain `Pipeline` instances. 

### New tests

*Relevant files*: `tests/test_voting_learners_api_*.py`, `tests/test_voting_learners_expts_*.py`, and `tests/utils.py`.

- The artificially constructed datasets we use for the existing tests are not very useful for `VotingLearner` tests since they are either too small or too toy-ish. Therefore, we use the digits and housing datasets (included in scikit-learn) for classification and regression tests respectively.  To make it easy to use these datasets, two new utility functions are added: `make_digits_data()` and `make_california_housing_data()`. 

- Note that we were already using a version of the digits dataset in the learning curve tests for the `Learner` class. This use was refactored to use the new `make_digits_data()` utility function. 

- All of the API tests test the methods of the `VotingLearner` class by first calling that method on an instantiated ``VotingLearner`` in SKLL space, then using only scikit-learn functions to perform (nearly) identical
operations, and then comparing the two results to make sure they are as close as possible. Most of the classification tests only those results up to 2 decimal places because there are some inherent differences between scikit-learn and SKLL that make it difficult to replicate SKLL operations in scikit-learn space. Most of the regression tests are able to compare to more decimal places since there are no probabilities involved. 

- Since the API tests are so comprehensive, there is no real reason for the experiment tests to run real experiments since the same API methods are called from within `run_configuration()` anyway. For this reason, we focus all of the experiment tests on making sure that the right methods are called based on the "task" value and that they are called with the right arguments derived from the fields specified in the configuration file. To do so, we mock the appropriate API methods, call `run_configuration()` on different configuration files, and check that the mocked methods were called the expected number of times and with the expected arguments. 

- To make experiment testing easier, we add a new utility function called `fill_in_config_options_for_voting_learners()` that takes an empty configuration file template (`tests/configs/test_voting_learner.template.cfg ) and populates it with the right values depending on the arguments with which the function was called. In addition, a new class called `BoolDict` is added that returns ``False`` as the default value for key lookups rather than ``None``. This simplifies the `fill_in_config_options_for_voting_learners()` function significantly. 

- New tests were added to `test_input.py` for the `save_votes` field. In addition, existing tests in the same file were updated to accommodate this new field.

- Note that there is still some code duplication between the tests but this adding them all into the same file will add a lot more complexity (`if` statements and the like). This way they are fully self-contained and can be run fully independently. 

### Documentation

*Relevant files*: ``doc/run_experiment.rst``, `doc/api/learner.rst`, ``examples/iris/voting.cfg``, ``examples/boston/voting.cfg``, and others (see below).

- Updated `run_experiment` documentation to include detailed description of `VotingClassifier` and `VotingRegressor` and added an entry for the `save_votes` configuration field. Note that some of the links will *only* work after the PR is merged. 
- Updated `doc/api/learner.rst` to include the `VotingLearner` class and improved sub-headings.
- Added new configuration files illustrating the use of `VotingClassifier` and `VotingRegressor` for the Iris and Boston examples respectively.
- Updated `doc/contributing.rst` page for readability and fixed links to existing methods. Note that some of the links will *only* work after the PR is merged. 
- Since top-level imports have been removed, deleted `api/skll.rst` and updated `api/quickstart.rst` and the Tutorial notebook. 
- Updated `doc/conf.py` to fix imports and changed year from 2019 to 2021. 

### Other changes

- There was a **major bug** when using samplers. We were [calling `fit_transform()`](https://github.com/EducationalTestingService/skll/blob/main/skll/learner/__init__.py#L1388) on the test set rather than `transform()`. This was fixed and `tests/test_classification.py:test_sampler:test_sparse_predict_sampler()` was updated.

- A new parallel build job was added to both Travis and Azure CI builds to accommodate the new tests. Test files were redistributed across all 6 jobs to make sure that the overall build time is still optimized. 

- The Travis CI configuration now creates a new conda environment rather than using the default miniconda one. In addition, it does not activate said environment when running the tests. Finally, it also configures `nosetests` via environment variables for simplicity. 

- Both Azure CI and Travis CI now set logging level for tests to `WARNING` and do not use `--nologcapture` which significantly reduces the size of the logs produced. 

- The warning and error messages in `Learner.learning_curve()` have been tweaked to be more concise.

### CODE REVIEW

File: skll/experiments/__init__.py [hunk 1]
```diff
@@ -356,55 +415,79 @@ def _classify_featureset(args):  # noqa: C901
...
             elif task == 'predict':
                 logger.info("Writing predictions")
                 # we set `class_labels` to `False` so that if the learner is
-                # probabilistic, probabilities are written instead of labels
+                # probabilistic, probabilities are written instead of labels;
+                # voting learners have an extra keyword argument indicating
+                # whether we want to save the individual predictions (or votes)
+                extra_kwargs = {}
+                if isinstance(learner, VotingLearner):
+                    extra_kwargs["individual_predictions"] = save_votes
```
Comment (mulhod) [inline skll/experiments/__init__.py:486]: RE: My earlier comment: Could this be done for xval too? Would it be as simple as adding an extra keyword argument?

File: skll/learner/__init__.py [hunk 2]
```diff
@@ -1385,7 +1355,7 @@ def predict(self,  # noqa: C901
                 self.logger.warning('SkewedChi2Sampler uses a dense matrix')
                 if sp.issparse(xtest):
                     xtest = xtest.todense()
-            xtest = self.sampler.fit_transform(xtest)
+            xtest = self.sampler.transform(xtest)
```
Comment (mulhod) [inline skll/learner/__init__.py:1358]: 👍

File: skll/learner/utils.py [hunk 3]
```diff
@@ -282,10 +285,11 @@ def compute_evaluation_metrics(metrics,
     # initialize a dictionary that will hold all of the metric scores
     metric_scores = {metric: None for metric in metrics}
 
-    # if we are a classifier and in probability mode, then
-    # `yhat` are probabilities so we need to compute the
-    # class indices separately and save them too
-    if probability and model_type == 'classifier':
+    # if we are doing classification and are a probablistic
+    # learner or a soft-voting meta learner, then `yhat` are
+    # probabilities so we need to compute the  class indices
```
Comment (mulhod) [inline skll/learner/utils.py:290]: If you are making other changes to this or other files:

```suggestion
    # probabilities so we need to compute the class indices
```

**Review by mulhod — CHANGES_REQUESTED** [review]
I have a couple questions and comments. (I have not done a full review yet, though. I just want to break it up since I won't get back to this till tomorrow probably.)

It looks really awesome so far! Exciting.

File: skll/experiments/__init__.py (hunk 1 above)
Comment (desilinguist) [reply skll/experiments/__init__.py:487]: Yes, it can! I just pushed new commits adding this functionality!

File: doc/run_experiment.rst [hunk 4]
```diff
@@ -1321,6 +1339,15 @@ Whether to save each of the K :ref:`model files <output_model_files>` trained du
 each step of a K-fold cross-validation experiment.
 Defaults to ``False``.
 
+save_votes *(Optional)*
+"""""""""""""""""""""""
+
+Whether to save the predictions from the individual estimators underlying a
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply doc/run_experiment.rst:1345]: Now it does.

File: skll/experiments/utils.py [hunk 5]
```diff
@@ -35,6 +37,20 @@ def default(self, obj):
         return json.JSONEncoder.default(self, obj)
 
 
+class PipelineTypeEncoder(json.JSONEncoder):
+    """
+    This class is used for serializing ``sklearn.pipline.Pipeline`` objects.
```
Comment (mulhod) [inline skll/experiments/utils.py:42]:
```suggestion
    This class is used for serializing ``sklearn.pipeline.Pipeline`` objects.
```

File: skll/learner/voting.py [hunk 6]
```diff
@@ -0,0 +1,961 @@
...
+              grid_search=True,
+              grid_objective=None,
+              grid_jobs=None,
+              shuffle=False):
+        """
+        Train the voting meta-estimator.
+
+        First, we train each of the underlying estimators (represented by
+        a skll ``Learner``), possibly with grid search. Then, we instantiate
+        a `VotingClassifier` or ``VotingRegressor`` as appropriate with the
```
Comment (mulhod) [inline skll/learner/voting.py:289]:
```suggestion
        a ``VotingClassifier`` or ``VotingRegressor`` as appropriate with the
```

File: skll/learner/voting.py [hunk 7]
```diff
@@ -0,0 +1,961 @@
...
+        scikit-learn ``Pipeline`` stored in the ``pipeline`` attribute
+        of each trained `Learner` instance as the estimator. Finally,
```
Comment (mulhod) [inline skll/learner/voting.py:291]:
```suggestion
        of each trained ``Learner`` instance as the estimator. Finally,
```

File: skll/learner/voting.py [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+
+        The trained meta-estimator is saved in the ``_model`` attribute.
+        Nothing is returned.
+
+        Parameters
+        ----------
+        examples : skll.FeatureSet
+            The ``FeatureSet`` instance to use for training.
+        param_grid_list : list, optional
+            The list of parameters grid to search through for grid
```
Comment (mulhod) [inline skll/learner/voting.py:307]:
```suggestion
            The list of parameter grids to search through for grid
```

File: skll/learner/voting.py [hunk 9]
```diff
@@ -0,0 +1,961 @@
...
+            search, one for each underlying learner. The order of
+            the dictionaries should correspond to the order If ``None``, the default
```
Comment (mulhod) [inline skll/learner/voting.py:309]: I think a sentence was perhaps not completed here?

File: skll/learner/voting.py [hunk 10]
```diff
@@ -0,0 +1,961 @@
...
+        else:
+            self._model_type = VotingRegressor
+            model_kwargs = {}
+        meta_learner = self.model_type(estimators, **model_kwargs)
+
+        # get the training features in the right dictionary format
+        X_train = examples.vectorizer.inverse_transform(examples.features)
+
+        # since label dictionaries are identical for all underlying
+        # learners, save it into a easier to access attribute
```
Comment (mulhod) [inline skll/learner/voting.py:371]:
```suggestion
        # learners, save it into an easier to access attribute
```

File: skll/learner/voting.py [hunk 11]
```diff
@@ -0,0 +1,961 @@
...
+        self.label_dict = self.learners[0].label_dict
+
+        # get the training label sin the right format too
```
Comment (mulhod) [inline skll/learner/voting.py:374]:
```suggestion
        # get the training labels in the right format too
```

File: skll/learner/voting.py [hunk 12]
```diff
@@ -0,0 +1,961 @@
...
+        if self.learner_type == "classifier":
+            y_train = [self.label_dict[label] for label in examples.labels]
+        else:
+            # for regressors, the labels are just the labels
+            y_train = examples.labels
+
+        # now we need to fit the actual meta learner which will also fit
+        # clones of the underlying pipelines;
+        # NOTE: this will *not* yield the same results as the case where we take
+        # the predictions from the trained SKLL learners above and did the
```
Comment (mulhod) [inline skll/learner/voting.py:387]:
```suggestion
        # the predictions from the trained SKLL learners above and do the
```

File: skll/learner/voting.py [hunk 13]
```diff
@@ -0,0 +1,961 @@
...
+         metric_scores) = compute_evaluation_metrics(output_metrics,
+                                                     ytest,
+                                                     yhat,
+                                                     self.learner_type,
+                                                     label_dict=train_and_test_label_dict,
+                                                     grid_objective=grid_objective,
+                                                     probability=self.voting == "soft",
+                                                     logger=self.logger)
+
+        # add in the model parameters,excluding the ones
```
Comment (mulhod) [inline skll/learner/voting.py:607]:
```suggestion
        # add in the model parameters, excluding the ones
```

File: tests/test_voting_learners_api_2.py [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+                          else VotingRegressor)
+    sklearn_model_kwargs = {"estimators": [('clf1', clf1),
+                                           ('clf2', clf2),
+                                           ('clf3', clf3)]}
+    if learner_type == "classifier":
+        sklearn_model_kwargs["voting"] = voting_type
+    sklearn_vl = sklearn_model_type(**sklearn_model_kwargs)
+    sklearn_vl.fit(train_fs.features, train_fs.labels)
+
+    # get the predictions from this voting classifier on the test ste
```
Comment (mulhod) [inline tests/test_voting_learners_api_2.py:127]:
```suggestion
    # get the predictions from this voting classifier on the test set
```

File: tests/test_voting_learners_api_3.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+    if with_individual_predictions:
+        for name, estimator in sklearn_vl.named_estimators_.items():
+            estimator_predictions = estimator.predict(test_fs.features)
+            # scikit-learn individual predictions are indices not class labels
+            # so we need to convert them to labels if required
+            if with_class_labels:
+                estimator_predictions = [sklearn_vl.classes_[index]
+                                         for index in estimator_predictions]
+            # if no class labels, then get the probabilities with soft
+            # voting; note that sinec the individual predictions from
```
Comment (mulhod) [inline tests/test_voting_learners_api_3.py:132]:
```suggestion
            # voting; note that since the individual predictions from
```

File: tests/test_voting_learners_api_3.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+            assert_array_equal(skll_max_prob_indices, sklearn_max_prob_indices)
+            # check individual probabilities but only for non-SVC estimators
+            if with_individual_predictions:
+                assert_array_almost_equal(skll_individual_dict["LogisticRegression"],
+                                          sklearn_individual_dict["LogisticRegression"],
+                                          decimal=2)
+                assert_array_almost_equal(skll_individual_dict["MultinomialNB"],
+                                          sklearn_individual_dict["MultinomialNB"],
+                                          decimal=2)
+        # in all other cases, we expect the actual class lables or class indices
```
Comment (mulhod) [inline tests/test_voting_learners_api_3.py:168]:
```suggestion
        # in all other cases, we expect the actual class labels or class indices
```

**Review by mulhod — CHANGES_REQUESTED** [review]
Minor typos.

**Review by mulhod — APPROVED** [review]
Really nice! Exciting changes!

**Review by aoifecahill — APPROVED** [review]
Looks great, thanks! I've tried it out (including navigating the documentation) and managed to successfully run and revise an experiment based on the outputs/documentation.
